hk_adc/hk_adc.c

```diff
@@ -0,0 +1,530 @@
+/*
+ * hk_adc.c
+ *
+ *  Created on: 05-Mar-2021
+ *      Author: Sonal Shrivastava
+ */
+
+#include "../firmware/drivers/mss_gpio/mss_gpio.h"
+#include "../firmware/drivers/mss_i2c/mss_i2c.h"
+#include "hk_adc.h"
+
+
+#define TX_LENGTH                       1u
+#define RX_LENGTH                       2u
+#define DUMMY_SERIAL_ADDR               0xFF
+#define NUM_SAMPLES_TO_AVG              10u
+
+
+/* Local Variables */
+static mss_i2c_status_t status;
+static uint16_t read_value;
+static uint8_t os_bit = 0;
+static uint8_t  rx_buffer[RX_LENGTH];
+static uint16_t volt_array[NUM_SAMPLES_TO_AVG];
+static uint16_t curr_array[NUM_SAMPLES_TO_AVG];
+static uint16_t citi_temp_array[NUM_SAMPLES_TO_AVG];
+static uint16_t hk_adc_avg_volt;
+static uint16_t hk_adc_avg_curr;
+static uint16_t hk_adc_avg_citi_temp;
+
+
+/* Local Function prototypes */
+static int hk_adc_reg_write(hk_adc_register_t reg, uint8_t *write_buffer);
+static int hk_adc_reg_read(hk_adc_register_t reg, uint16_t *read_buffer);
+static void hk_adc_update_multiplier_volt(void);
+static uint16_t hk_adc_calc_avg_voltage(void);
+static uint16_t hk_adc_calc_avg_current(void);
+static uint16_t hk_adc_calc_avg_citi_temp(void);
+
+
+/**
+ * @brief HK ADC Init function
+ *
+ * This function is called once to initialize ADC with a certain configuration
+ * - FSR - +-4.096, Single-shot conversion, DR = 1600SPS, MUX input CITI_TEMP and
+ * disabled comparator.
+ *
+ *
+ * @param none.
+ * @return HK_ADC_ERR_INIT_FAILED if any error occurs otherwise HK_ADC_NO_ERR.
+ */
+int hk_adc_init(void)
+{
+    uint8_t err = HK_ADC_ERR_INIT_FAILED;
+
+    // By default, ADC is configured to point towards CITI_TEMP as MUX input
+    uint16_t config = HK_ADC_CONFIG_MUX_SINGLE_2 | HK_ADC_CONFIG_FSR_2 | HK_ADC_CONFIG_MODE_SINGLE_SHOT | HK_ADC_CONFIG_DR_1600SPS | HK_ADC_CONFIG_CMODE_TRAD | HK_ADC_CONFIG_CPOL_ACTIVE_LOW | HK_ADC_CONFIG_CLAT_NON_LATCH | HK_ADC_CONFIG_CQUE_DISABLE;
+    uint8_t send_buffer[2] = {config>>8, config};
+
+    // Set the clock for i2c0 instance
+    MSS_I2C_init(&g_mss_i2c0, DUMMY_SERIAL_ADDR, MSS_I2C_PCLK_DIV_256); //10^8/256 = 390,625Hz =~ 390KHz
+
+    // Write to the CONFIG register
+    err = hk_adc_reg_write(HK_ADC_REG_CONFIG, &send_buffer[0]);
+
+    if (err == HK_ADC_NO_ERR)
+    {
+        return err;
+    }
+
+    err = HK_ADC_ERR_INIT_FAILED;
+    return err;
+}
+
+
+
+
+/**
+ * @brief HK ADC read Battery Voltage function
+ *
+ * This function returns the raw battery voltage value as read by the ADC. ADC
```

Comment:
I would rephrase "returns" here, to avoid confusion in the future.

Reply:
Made the change. Hope it looks better now.